ls1intum/Artemis
# `Programming exercises`: Remove duplicate alerts and fix initial dates on page load

### Checklist
#### General


#### Client

### Motivation and Context


Addresses duplicate alert issue mentioned in #5248 and fixes related issues in `ProgrammingExerciseLifecycleComponent` described in the next section.

### Description

Closes #5248. Additionally:
- Prevents auto-modifications to dates in read-only mode for programming exercises.
- Fixes initial dates on page load. Previously, after displaying alerts like `The example solution publication date has been set to the due date`, the dates were not set to correct values, causing saving to fail without manual user intervention.
- Fixes the source of the new date (can be release date or due date) in the alert message when auto-setting the example solution publication date. Tries to minimize number of subsequent alerts for date fixes on edit page load.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with invalid release date, due date and example solution publication date combinations. (You shouldn't be able to create new invalid ones due to server-side validations so please **do not save the invalid exercises** you find) You can use https://artemis-test1.artemis.in.tum.de/course-management/18/programming-exercises/107/ for testing without saving.

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to details page of the invalid programming exercise and make sure you do not see any browser alerts about invalid dates.
4. Click edit, now you should only see each alert message mentioning invalid date set to a new value at most once before manually changing the dates.
5. Set release date and due date to dates later than example solution publication date, alert message should mention the source of new date correctly.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| programming-exercise-lifecycle.component.ts| 87.5% | 95.65% |
